llpc/lower/llpcSpirvLowerReverseThreadGroup.cpp

```diff
@@ -0,0 +1,176 @@
+
+// =====================================================================================================================
+// Pass creator, creates the pass of SPIR-V lowering operations for reversing thread group
+ModulePass *createLegacySpirvLowerReverseThreadGroup() {
+  return new LegacySpirvLowerReverseThreadGroup();
+}
+
+// =====================================================================================================================
+LegacySpirvLowerReverseThreadGroup::LegacySpirvLowerReverseThreadGroup() : ModulePass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+bool LegacySpirvLowerReverseThreadGroup::runOnModule(Module &module) {
+  return Impl.runImpl(module);
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses SpirvLowerReverseThreadGroup::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  runImpl(module);
+  return PreservedAnalyses::none();
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool SpirvLowerReverseThreadGroup::runImpl(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass Spirv-Lower-Reverse-Thread-Group\n");
+
+  SpirvLower::init(&module);
+
+  if (m_shaderStage != ShaderStageCompute)
+    return false;
+
+  visit(module);
+
+  processBuiltIns();
+
+  return m_workgroupId || m_globalInvocationId;
+}
+
+} // namespace Llpc
+
+// =====================================================================================================================
+// Visits "call" instruction.
+// Looks for @lgc.create.read.builtin.input.* calls for WorkgroupId and GlobalInvocationId, and records the built-in
+// value.
+//
+// @param callInst : "Call" instruction
+void SpirvLowerReverseThreadGroup::visitCallInst(CallInst &callInst) {
+  auto callee = callInst.getCalledFunction();
+  if (!callee)
+    return;
+
+  auto mangledName = callee->getName();
+  if (!mangledName.startswith("lgc.create.read.builtin.input"))
+    return;
+
+  unsigned builtIn = cast<ConstantInt>(callInst.getOperand(0))->getZExtValue();
+
+  if (builtIn == lgc::BuiltInWorkgroupId) {
+    assert(!m_workgroupId);
+    m_workgroupId = &callInst;
+  } else if (builtIn == lgc::BuiltInGlobalInvocationId) {
+    assert(!m_globalInvocationId);
+    m_globalInvocationId = &callInst;
+  }
```

Comment:
Can these intrinsics only get called once?

Reply:
I believe SpirvLowerGlobal::mapInputToProxy() will turn all uses of the builtins into proxies, and the @lgc.create.read.builtin.input.* call will only appear once when initializing one proxy.